src/nnet3/nnet-general-component.cc

```diff
@@ -880,6 +880,263 @@ void StatisticsPoolingComponent::Backprop(
+         << ", zeroing-threshold=" << zeroing_threshold_
+         << ", zeroed-proportion="
+         << (count_zeroing_boundaries_ > 0 ?
+             static_cast<BaseFloat>(num_zeroed_)/count_zeroing_boundaries_ : 0);
+  return stream.str();
+}
+
+void BackpropTruncationComponent::Init(int32 dim,
+                                 BaseFloat clipping_threshold,
+                                 BaseFloat zeroing_threshold,
+                                 int32 zeroing_interval,
+                                 int32 recurrence_interval,
+                                 int32 num_clipped,
+                                 int32 num_zeroed,
+                                 int32 count,
+                                 int32 count_zeroing_boundaries) {
+  KALDI_ASSERT(clipping_threshold >= 0 && zeroing_threshold >= 0 &&
+      zeroing_interval > 0 && recurrence_interval > 0 && dim > 0);
+  dim_ = dim;
+  clipping_threshold_ = clipping_threshold;
+  zeroing_threshold_ = zeroing_threshold;
+  zeroing_interval_ = zeroing_interval;
+  recurrence_interval_ = recurrence_interval;
+  num_clipped_ = num_clipped;
+  num_zeroed_ = num_zeroed;
+  count_ = count;
+  count_zeroing_boundaries_ = count_zeroing_boundaries;
+}
+
+// virtual
+void BackpropTruncationComponent::InitFromConfig(ConfigLine *cfl) {
+  int32 dim = 0;
+  bool ok = cfl->GetValue("dim", &dim);
+  BaseFloat clipping_threshold = 15.0;
+  BaseFloat zeroing_threshold = 2.0;
+  int32 zeroing_interval = 20, recurrence_interval = 1;
+  cfl->GetValue("clipping-threshold", &clipping_threshold);
+  cfl->GetValue("zeroing-threshold", &zeroing_threshold);
+  cfl->GetValue("zeroing-interval", &zeroing_interval);
+  cfl->GetValue("recurrence-interval", &recurrence_interval);
+  if (!ok || cfl->HasUnusedValues() ||
+      clipping_threshold < 0 || zeroing_threshold < 0 || zeroing_interval < 1 ||
+      recurrence_interval < 1 || dim <= 0)
+    KALDI_ERR << "Invalid initializer for layer of type "
+              << Type() << ": \"" << cfl->WholeLine() << "\"";
+  Init(dim, clipping_threshold, zeroing_threshold,
+       zeroing_interval, recurrence_interval, 0, 0, 0, 0);
+}
+
+// virtual
+ComponentPrecomputedIndexes*
+BackpropTruncationComponent::PrecomputeIndexes(
+    const MiscComputationInfo &misc_info,
+    const std::vector<Index> &input_indexes,
+    const std::vector<Index> &output_indexes,
+    bool need_backprop) const {
+  int32 num_input_indexes = input_indexes.size(),
+      num_output_indexes = output_indexes.size();
+  KALDI_ASSERT(num_input_indexes == num_output_indexes);
+  BackpropTruncationComponentPrecomputedIndexes *ans = new
+      BackpropTruncationComponentPrecomputedIndexes();
+  ans->zeroing.Resize(num_output_indexes);
+
+  for (int32 i = 0; i < num_output_indexes; i++) {
+    const int32 output_n = output_indexes[i].n;
+    const int32 output_t = output_indexes[i].t;
+    if ((output_t - output_n % zeroing_interval_) / zeroing_interval_ == 
```

Comment:
Also, instead of  if (x == y + 1), do if ( x != y) ... this will generalize better to weird situations like when zeroing_interval_ is smaller than recurrence_interval_.